source/log.c

```diff
@@ -0,0 +1,386 @@
+             aws_thread_current_thread_id(), fmt);
+
+    int count = vsnprintf(msg_data, s_local_log_context->max_message_len, fmt_final, va_args);
+    if ((size_t)count >= s_local_log_context->max_message_len - 1) {
+        msg_data[s_local_log_context->max_message_len - 2] = '\n';
+    }
+
+    /* Push formatted message onto this thread's message list. Will be picked up later and
+    processed by the logging thread in `aws_log_flush()`, and handed to a user callback set by
+    `aws_log_set_reporting_callback`. */
+    AWS_ATOMIC_SINGLY_LIST_INSERT(s_local_log_context->message_list, msg);
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_log(enum aws_log_level level, const char *fmt, ...) {
+    if (!s_local_log_context || !s_log_system_running) {
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    va_list va_args;
+    va_start(va_args, fmt);
+    int ret = aws_vlog(level, fmt, va_args);
+    va_end(va_args);
+    return ret;
+}
+
+const char *aws_log_level_to_string(enum aws_log_level level) {
+    switch (level) {
+        case AWS_LOG_LEVEL_OFF:   return "OFF";
+        case AWS_LOG_LEVEL_FATAL: return "FATAL";
+        case AWS_LOG_LEVEL_ERROR: return "ERROR";
+        case AWS_LOG_LEVEL_WARN:  return "WARN";
+        case AWS_LOG_LEVEL_TRACE: return "TRACE";
+        default: return NULL;
+    }
+}
+
+static void s_aws_log_remove_dead_context(int index) {
+    assert(index >= 0 && index < AWS_LOG_MAX_LOGGING_THREADS);
+
+    /* Atomically grab the ctx pointer for this index. */
+    struct aws_log_context* ctx = aws_atomic_store_ptr((void **)&s_log_table[index].ctx, NULL);
+
+    if (ctx) {
+        /* Cleanup all resources for this context. */
+        struct aws_log_message *msg_list = ctx->message_list;
+        AWS_SINGLY_LIST_REVERSE(struct aws_log_message, msg_list);
+        int count = 0;
+
+        while (msg_list) {
+            struct aws_log_message *next0 = msg_list->next;
+            char *msg_data = (char*)(msg_list + 1);
+            s_log_report_callback(msg_data);
+            aws_memory_pool_release(ctx->message_pool, msg_list);
+            msg_list = next0;
+            ++count;
+        }
+
+        aws_atomic_add(&s_log_message_count, -count);
+
+        struct aws_log_message *delete_list = ctx->delete_list;
+        while (delete_list) {
+            struct aws_log_message *next0 = delete_list->next;
+            aws_memory_pool_release(ctx->message_pool, delete_list);
+            delete_list = next0;
+        }
+
+        void *arena;
+        size_t arena_size;
+        aws_memory_pool_get_arena(ctx->message_pool, &arena, &arena_size);
+        aws_secure_zero(arena, arena_size);
+        aws_memory_pool_clean_up(ctx->message_pool);
+        aws_mem_release(ctx->alloc, ctx);
+    }
+
+    int delete_me = AWS_LOG_ENTRY_STATE_DELETEME;
+    if (aws_atomic_compare_exchange(&s_log_table[index].state, &delete_me, AWS_LOG_ENTRY_STATE_NO_WRITERS)) {

```

Comment:
This means the thread's log context was marked for deletion while this code path was being executed. As this code path is coming to a close, it will perform the cleanup function (it is responding to the `DELETEME` "signal".